### ISSUE TITLE
Fix pyflakes for conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7002]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7002|
|Type|enhancement|
|Created|2014-02-27 13:56:03Z|
|Branch|https://github.com/twisted/twisted/tree/pyflakes-for-conch-7002|

Attachments:

* [7002-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/760/76096e8f6bea09a68de0d6716c47896a42ad6fc2/d8e7a8b025c7e56dfa02ee7f07070b14c6885290.diff) (14155 bytes) - added by adiroiban on 2014-02-27 14:18:13Z - 
* [7002-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/760/76096e8f6bea09a68de0d6716c47896a42ad6fc2/1f28ab86916f69807df7ceec879407241df2dd23.diff) (15367 bytes) - added by adiroiban on 2014-03-03 08:59:35Z - 
* [7002-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/760/76096e8f6bea09a68de0d6716c47896a42ad6fc2/506fc0a02fe6047c2ea17b915ed064bb7aacaee6.diff) (17585 bytes) - added by adiroiban on 2014-03-17 20:03:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7002 7002
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_pyflakes_for_conch_7002 branches/pyflakes-for-conch-7002
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1393509363000000 1393509363000000
changetime__1397687810000000 1397687810000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

twisted/conch/scripts/cftp.py:431: undefined name 'files' is still not fixed as I don't know how to fix it.

twisted/conch/ssh/filetransfer.py:58: undefined name 'e' was fixed by definining Exception, e. It looks like an existing bug and maybe this need a separate ticket so that a test is written for this code.


So only one outstanding problem not fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @adiroiban|
|-|-|

Thanks.

  1. `twisted/conch/manhole_ssh.py` - removing `session = self.proto.session` is not correct.  The pyflakes warning is _redefinition of unused 'session' from line 14_.  Removing this line causes `sendClose` to be called with the wrong argument.  Happily this has caused some failures in `twisted.conch.test.test_manhole` :)
  1. `twisted/conch/scripts/ckeygen.py`
    1. It's usually desirable to avoid a top-level reactor import in library code.  Since the first reactor import installs the default reactor this can interfere with an application's choice of reactor.
    1. Fortunately in this case I think the reactor import can be completely eliminated.  It's only used by `handleError` to stop the reactor on error.  And this makes no sense because ckeygen never runs the reactor.  I think `handleError` and the code that uses it should be deleted.  As the code that uses it say, anyway, it is a hack.
  1. `twisted/conch/ssh/agent.py` - This change is also wrong.  It's another _redefinition of unused ..._ warning.
  1. `twisted/conch/ssh/filetransfer.py`
    1. Please use the new exception syntax for all contributions to Twisted - `except Exception as e:`
    1. But this exception block looks like a big mess.  `e` isn't actually used because of the `continue` statement halfway through...  I think the fix here is to leave `except` alone and delete the lines from `continue` onward.  The lines after `continue` are clearly dead code and once they're gone this exception handler is the end of the loop so the `continue` itself is redundant.
  1. twisted/conch/test/test_agent.py - I dislike adding useless code to placate stupid tools.  There are a number of possible better fixes here (perhaps any and all of which are better deferred to a follow-up ticket):
    1. Introduce a `requireModule("OpenSSL")` style API so we don't have to write code that pyflakes is sure is useless.  This is both more expressive to pyflakes and to future maintainers of the code.
    1. Fix `iosim` so that it only declares it implements `ITLSTransport` if pyOpenSSL is available and then check this declaration in `test_agent.py` instead of relying on `ImportError`.
    1. Check the reactor for TLS support (since its declaration of `IReactorSSL` is already conditional).
  1. `twisted/conch/test/test_checkers.py`, `twisted/conch/test/test_ckeygen.py`, `twisted/conch/test/test_default.py`, `twisted/conch/test/test_endpoints.py`, `twisted/conch/test/test_knownhosts.py`, `twisted/conch/test/test_openssh_compat.py`, `twisted/conch/test/test_scripts.py`, `twisted/conch/test/test_tap.py`, `twisted/conch/test/test_userauth.py` - ditto
  1. Sort of ditto for `twisted/conch/test/test_insults.py`, but using the existing `namedAny` for this kind of usage seems like it makes sense to me.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @exarkun|
|-|-|

Thanks for the review.

Here are my comments.

Patched attached.

Conch test pass on my computer in 2.7.

Hope that all problems were addressed.


  1. `twisted/conch/manhole_ssh.py` - removing `session =
 self.proto.session` is not correct.  The pyflakes warning is
 _redefinition of unused 'session' from line 14_.  Removing this line
 causes `sendClose` to be called with the wrong argument.  Happily this has
 caused some failures in `twisted.conch.test.test_manhole` :)

True. My bad. Happy to see that test are doing their job.


   1. `twisted/conch/scripts/ckeygen.py`
     1. It's usually desirable to avoid a top-level reactor import in
 library code.  Since the first reactor import installs the default reactor
 this can interfere with an application's choice of reactor.
     1. Fortunately in this case I think the reactor import can be
 completely eliminated.  It's only used by `handleError` to stop the
 reactor on error.  And this makes no sense because ckeygen never runs the
 reactor.  I think `handleError` and the code that uses it should be
 deleted.  As the code that uses it say, anyway, it is a hack.


Done that. Thanks!

   1. `twisted/conch/ssh/agent.py` - This change is also wrong.  It's
 another _redefinition of unused ..._ warning.

Thanks. Should be fixed.

   1. `twisted/conch/ssh/filetransfer.py`
     1. Please use the new exception syntax for all contributions to
 Twisted - `except Exception as e:`
     1. But this exception block looks like a big mess.  `e` isn't actually
 used because of the `continue` statement halfway through...  I think the
 fix here is to leave `except` alone and delete the lines from `continue`
 onward.  The lines after `continue` are clearly dead code and once they're
 gone this exception handler is the end of the loop so the `continue`
 itself is redundant.

Done. Removed unused code.

   1. twisted/conch/test/test_agent.py - I dislike adding useless code to
 placate stupid tools.  There are a number of possible better fixes here
 (perhaps any and all of which are better deferred to a follow-up ticket):
     1. Introduce a `requireModule("OpenSSL")` style API so we don't have
 to write code that pyflakes is sure is useless.  This is both more
 expressive to pyflakes and to future maintainers of the code.

Done. [#7014](https://github.com/twisted/twisted/issues/7014)

     1. Fix `iosim` so that it only declares it implements `ITLSTransport`
 if pyOpenSSL is available and then check this declaration in
 `test_agent.py` instead of relying on `ImportError`.

Done. [#7015](https://github.com/twisted/twisted/issues/7015)

     1. Check the reactor for TLS support (since its declaration of
 `IReactorSSL` is already conditional).

Hope is done. Here is what I understood and implemented:

```
from twisted.internet import reactor
from twisted.internet.interfaces import ITLSTransport

if ITLSTransport.providedBy(reactor):
    from twisted.test import iosim
else:
    iosim = None
```


   1. `twisted/conch/test/test_checkers.py`,
 `twisted/conch/test/test_ckeygen.py`,
 `twisted/conch/test/test_default.py`,
 `twisted/conch/test/test_endpoints.py`,
 `twisted/conch/test/test_knownhosts.py`,
 `twisted/conch/test/test_openssh_compat.py`,
 `twisted/conch/test/test_scripts.py`, `twisted/conch/test/test_tap.py`,
 `twisted/conch/test/test_userauth.py` - ditto

I placed # FIXME: [#7014](https://github.com/twisted/twisted/issues/7014) comments to clean the code once new feature
is implemented.

   1. Sort of ditto for `twisted/conch/test/test_insults.py`, but using the
 existing `namedAny` for this kind of usage seems like it makes sense to
 me.

I have implemented the code using namedAny

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @adiroiban|
|-|-|

Apart from the [#7014](https://github.com/twisted/twisted/issues/7014) issues remaining this looks pretty good now.  I notice there are still pyflakes errors in some Twisted Conch modules, though.  It looks like this patch fixes all of the import-related warnings where the imported module isn't directly used.  Some other warnings come from the fact that code that runs later wants to notice whether the module could be imported or not.

Perhaps `namedModule` could address both of these issues at once - if it were extended to support a default result.

```
try:
    namedModule("pyasn1")
except ImportError as e:
    pyasn1Skip = str(e)
else:
    pyasn1Skip = None
```

or 

```
DES3 = namedModule("Crypto.Cipher.DES3", default=None)
```

Or maybe the need to extend `namedModule` to handle both these cases really does suggest that there should be a dedicated helper like `requireModule` that can add this new optional default behavior.

Either way I think this patch is great apart from the [[#7014](https://github.com/twisted/twisted/issues/7014)](https://github.com/twisted/twisted/issues/7014) issue.  We could apply it now if you'd like to deal with the remaining issues in a follow-up ticket or we can wait until #7014 is resolved - whichever you prefer.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @exarkun|
|-|-|

Hi,

Thanks for the review.

What other errors do you get? Do you have the latest pyflakes?

Here is what I got on my computer:

```
$ pyflakes --version
0.7.3
$ pyflakes twisted/conch/
twisted/conch/scripts/cftp.py:431: undefined name 'files'
```

There is still an error which I would like to fix before merge, but I need help and I have no idea what is going on there... maybe this needs a separate ticket!


```

    def _cbPutTargetAttrs(self, attrs, path, local):
        if not stat.S_ISDIR(attrs['permissions']):
            return "Wildcard put with non-directory target."
        return self._cbPutMultipleNext(None, **files**, path)

```

-------

> Or maybe the need to extend namedModule to handle both these cases really does suggest that there should be a dedicated helper like requireModule that can add this new optional default behavior.

I think that we should go for requireModule('string', default=None) API, and not use the existing namedModule method.

I find that namedModule method name is strange... and as you have seen, it peoples forget about it.

>
> Either way I think this patch is great apart from the [[#7014](https://github.com/twisted/twisted/issues/7014)](https://github.com/twisted/twisted/issues/7014) issue. 

I will submit a fix for #7014 .

Please suggest a fix for cftp.py:431 

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Please suggest a fix for cftp.py:431

Offhand I can't see what the correct fix for this error is.  If this code path ever ran it would fail by raising a `NameError`.  This strongly suggests the code path is never run by the test suite - which in turn suggests that the best thing to do next for this code is write a unit test that exercises this code.  Hopefully once such a test is written someone will have some idea what the point of this code is and then it may be easier to determine the correct fix.

I suggest dealing with this error separately so that the rest of these great fixes can be merged before someone figures out how to test and fix this part of cftp.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

Created a new ticket and will look into that [#7037](https://github.com/twisted/twisted/issues/7037) . I really want to have clean pyflakes results.

Apart from news file, is there anything wrong this the last patch?

If not, can you please create a news file and merge this?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @adiroiban|
|-|-|

Ah.... requireModule API landed on trunk. I will update the code and also include the newsfile. Sorry for the noise.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @exarkun|
|-|-|

Hi,
It looked like nameAny was enough to fix pyflakes errors.

I have attached a new diff based on latest trunk.

It has a topfile.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [42409]) Branching to pyflakes-for-conch-7002.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

Thanks for your tireless work on this adiroiban!

I've merged your patch, all your changes look good!  I made a couple of small changes:
 1.  There still seems to be one more pyflakes complaint:  twisted/conch/ui/ansi.py:148: local variable 'mode' is assigned to but never used (I am on pyflakes version 0.8.1), so I fixed that
 1.  Fixed some twisted checker whitespace issues (there are more, they seem to be problems with twisted checker)
 1.  I converted namedAny to requireModule for some of the files.

All tests pass locally, and buildbot passes.  Merging.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli set status to `closed`|
|-|-|

(In [42428]) Merge pyflakes-for-conch-7002

Author: adiroiban
Reviewer: exarkun, cyli
Fixes: [#7002](https://github.com/twisted/twisted/issues/7002)

Fix most pyflakes errors in the conch module.
